nubis/bin/migrate

```diff
@@ -1,34 +1,119 @@
 #!/bin/bash
+#
+# This script is run on every ami change.
+#+ This is the place to do things like database initilizations and migrations.
+#
+#set -x
+
+LOGGER_BIN='/usr/bin/logger'
+
+# Set up the logger command if the binary is installed
+if [ ! -x $LOGGER_BIN ]; then
+    echo "ERROR: 'logger' binary not found - Aborting"
+    echo "ERROR: '$BASH_SOURCE' Line: '$LINENO'"
+    exit 2
+else
+    LOGGER="$LOGGER_BIN --stderr --priority local7.info --tag migrate.sh"
+fi
+
+# Source the consul connection details from the metadata api
 eval `ec2metadata --user-data`
 
-echo "Environment: $NUBIS_ENVIRONMENT"
-echo "Project: $NUBIS_PROJECT"
+# Check to see if NUBIS_MIGRATE was set in userdata. If not we exit quietly.
+if [ ${NUBIS_MIGRATE:-0} == '0' ]; then
+    exit 0
+fi
 
+# Set up the consul url
 CONSUL="http://localhost:8500/v1/kv/$NUBIS_PROJECT/$NUBIS_ENVIRONMENT/config"
 
 # We run early, so we need to account for Consul's startup time, unfortunately, magic isn't
 # always free
-
 CONSUL_UP=-1
+COUNT=0
 while [ "$CONSUL_UP" != "0" ]; do
-  QUERY=`curl -s http://localhost:8500/v1/kv/$NUBIS_PROJECT/$NUBIS_ENVIRONMENT/config?raw=1`
-  CONSUL_UP=$?
+    if [ ${COUNT} == "6" ]; then
+        $LOGGER "ERROR: Timeout while attempting to connect to consul."
+        exit 1
+    fi
+    QUERY=`curl -s ${CONSUL}?raw=1`
+    CONSUL_UP=$?
 
-  if [ "$QUERY" != "" ]; then
-    CONSUL_UP=-2
-  fi
+    if [ "$QUERY" != "" ]; then
+        CONSUL_UP=-2
+    fi
 
-  echo "Consul not ready yet ($CONSUL_UP), retrying..."
-  sleep 1
+    if [ "$CONSUL_UP" != "0" ]; then
+      $LOGGER "Consul not ready yet ($CONSUL_UP). Sleeping 10 seconds before retrying..."
+      sleep 10
+      COUNT=${COUNT}+1
+    fi
 done
 
+# Generate and set the secrets for the app
+DB_NAME=`curl -s $CONSUL/DB_NAME?raw=1`
+if [ "$DB_NAME" == "" ]; then
+    DB_NAME='dpaste'
+    curl -s -X PUT -d $DB_NAME $CONSUL/DB_NAME
+fi
+$LOGGER " + DB_NAME=$DB_NAME"
+
+DB_USERNAME=`curl -s $CONSUL/DB_USERNAME?raw=1`
+if [ "$DB_USERNAME" == "" ]; then
+    DB_USERNAME='dpaste'
+    curl -s -X PUT -d $DB_USERNAME $CONSUL/DB_USERNAME
+fi
+$LOGGER " + DB_USERNAME=$DB_USERNAME"
 
-APP_SECRET_KEY=`curl -s $CONSUL/app_secret_key?raw=1`
+DB_PASSWORD=`curl -s $CONSUL/DB_PASSWORD?raw=1`
```

Comment:
If the DB_PASSWORD contains a single-quote character ```'```, the script will break when trying to make use of that password later on. It may be desirable to refuse passwords that contain ```'```.

(We've seen these before in prod from users who handcrafted a 'complex' password on their own, and also in passphrases that contain English contractions.)